code/espurna/curtain_kingart.cpp

```diff
@@ -87,108 +235,248 @@ To configure the device:
-void _KACurtainSetclose(const char * message) {
-    char tx_buffer[80] = {0};
-    snprintf_P(
-        tx_buffer, sizeof(tx_buffer),
-        PSTR("AT+UPDATE=\"sequence\":\"%d%03u\",\"switch\":\"%s\",\"setclose\":%s"),
-        now(), millis() % 1000,
-        "off", message
-    );
-    _KACurtainSend(tx_buffer);
-}
-  
-#if MQTT_SUPPORT
 
-void _KACurtainActionSelect(const char * message) {
-    if (strcmp(message, "pause") == 0) {
-        _KACurtainPause(message);
-    } else {
-        _KACurtainSetclose(message);
+    //Reset last button to make the algorithm work and set last button state
+    if(!_curtain_moving) 
+        _curtain_last_button = CURTAIN_BUTTON_UNKNOWN;
+    else if (_curtain_button != CURTAIN_BUTTON_UNKNOWN)
+        _curtain_last_button = _curtain_button;
+    
+    // Handle configuration button presses
+    if (buffer.indexOf("enterESPTOUCH") > 0) {
+        wifiStartAP();
+    } else if (buffer.indexOf("exitESPTOUCH") > 0) {
+        deferredReset(100, CUSTOM_RESET_HARDWARE);
+    } else { //In any other case, update as it could be a move action
+        curtainUpdateUI();
     }
 }
 
-void _KACurtainCallback(unsigned int type, const char * topic, char * payload) {
+// -----------------------------------------------------------------------------
+// MQTT Support
+// -----------------------------------------------------------------------------
+
+#if MQTT_SUPPORT
+
+//------------------------------------------------------------------------------
+void _curtainMQTTCallback(unsigned int type, const char * topic, char * payload) {
     if (type == MQTT_CONNECT_EVENT) {
         mqttSubscribe(MQTT_TOPIC_CURTAIN);
-    }
-    if (type == MQTT_MESSAGE_EVENT) {
+    } else if (type == MQTT_MESSAGE_EVENT) {
         // Match topic
         const String t = mqttMagnitude(const_cast<char*>(topic));
         if (t.equals(MQTT_TOPIC_CURTAIN)) {
-            _KACurtainActionSelect(payload);
+            if (strcmp(payload, "pause") == 0)
+                _KACurtainSet(CURTAIN_BUTTON_PAUSE);
+            else  if (strcmp(payload, "on") == 0)
+                _KACurtainSet(CURTAIN_BUTTON_OPEN);
+            else  if (strcmp(payload, "off") == 0)
+                _KACurtainSet(CURTAIN_BUTTON_CLOSE);
+            else {
+                _curtain_position_set = String(payload).toInt();
+                _KACurtainSet(CURTAIN_BUTTON_UNKNOWN, _curtain_position_set);
+                
+            }
         }
     }
 }
 
 #endif // MQTT_SUPPORT
-  
+
+// -----------------------------------------------------------------------------
+// WEB Support
+// -----------------------------------------------------------------------------
+
+#if WEB_SUPPORT
+
+//------------------------------------------------------------------------------
+void _curtainWebSocketOnConnected(JsonObject& root) {
+    root["curtainType"] = getSetting("curtainType", "0");
+    root["curtainInitialBehaviour"] = getSetting("curtainInitialBehaviour", "0");
```

Comment:
`curtainInitialBehaviour` -> `curtainBoot`?
"0" is only explained in web

Reply:
curtainBt/curtainBtPos